packages/HilltopCrawler/src/main/kotlin/nz/govt/eop/hilltop_crawler/api/HilltopFetcher.kt

```diff
@@ -0,0 +1,40 @@
+package nz.govt.eop.hilltop_crawler.api
+
+import io.github.bucket4j.Bandwidth
+import io.github.bucket4j.BlockingBucket
+import io.github.bucket4j.Bucket
+import java.net.URI
+import java.time.Duration
+import mu.KotlinLogging
+import org.springframework.stereotype.Component
+import org.springframework.web.client.RestClientException
+import org.springframework.web.client.RestTemplate
+
+/**
+ * Simple wrapper around RestTemplate to limit the number of requests per second This is to avoid
+ * overloading the Hilltop servers
+ *
+ * Could be extended to have a bucket per host to increase throughput. But currently the bottleneck
+ * is when initially loading data from Hilltop, which ends up effectively processing one host at a
+ * time.
+ */
+@Component
+class HilltopFetcher(val restTemplate: RestTemplate) {
+
+  private final val logger = KotlinLogging.logger {}
+
+  private final val bucketBandwidthLimit: Bandwidth = Bandwidth.simple(20, Duration.ofSeconds(1))
+  private final val bucket: BlockingBucket =
+      Bucket.builder().addLimit(bucketBandwidthLimit).build().asBlocking()
+
+  fun fetch(fetchRequest: URI): String? {
+    bucket.consume(1)
+    logger.trace { "Downloading $fetchRequest" }
+    return try {
+      restTemplate.getForObject(fetchRequest, String::class.java)
+    } catch (e: RestClientException) {
+      logger.info(e) { "Failed to download $fetchRequest" }
+      null
```

Comment:
Different failure reasons will cause the next attempt to be queued at different intervals, so the `runTask` function in the fetcher has code to handle different things going wrong, and an overall exception handler to try and ensure every task will be requeued in case of error rather than looping.

This particular example is just returning `null` to indicate that we failed to fetch ... with the assumption it was a server error, so that `runTask` can queue it again for soon-ish

Reply:
Cool, makes sense. My question was more about the HillTop crawler as a whole, but maybe that's better as a discussion.

With return values that have a meaning such as this, might be worth considering simple `Result` DataClasses class at some point, to make the code a bit more readable.